Parens on union type fail for function arguments (JSDoc style)

<!-- 🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨

Half of all issues filed here are duplicates, answered in the FAQ, or not appropriate for the bug tracker. Even if you think you've found a *bug*, please read the FAQ first, especially the Common "Bugs" That Aren't Bugs section!

Please help us by doing the following steps before logging an issue:
  * Search: https://github.com/Microsoft/TypeScript/search?type=Issues
  * Read the FAQ: https://github.com/Microsoft/TypeScript/wiki/FAQ

Please fill in the *entire* template below.
-->

<!-- Please try to reproduce the issue with `typescript@next`. It may have already been fixed. -->
**TypeScript Version:**  typescript@3.1.0-dev.20180717

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:** function arguments grouped union type

**Code**

I notice that if a union type is grouped with `(` and `)` it works as a return type but not as an argument type.  For example, this does not work:

```js
/**
 * @typedef {function((string|undefined)): (string|undefined)} SomeType
 */
```

While this does work:

```js
/**
 * @typedef {function(string|undefined): (string|undefined)} SomeType
 */
```

**Expected behavior:**

With this `tsconfig.json`:
```json
{
  "compilerOptions": {
    "target": "ES2017",
    "allowJs": true,
    "checkJs": true,
    "noEmit": true,
    "strict": false,
  },
  "include": [
    "src/**/*.js"
  ]
}
```

I expected `tsc` to succeed.

**Actual behavior:**

```

src/main.js:2:23 - error TS1138: Parameter declaration expected.

2  * @typedef {function((string|undefined)): (string|undefined)} SomeType
                        ~

src/main.js:2:41 - error TS1005: '}' expected.

2  * @typedef {function((string|undefined)): (string|undefined)} SomeType
                                          ~
```

Perhaps this is a known limitation.  I can remove the parens.  But I thought it might also be a parsing bug.

--- a/src/compiler/parser.ts
+++ b/src/compiler/parser.ts
@@ -1509,7 +1509,9 @@ namespace ts {
                 case ParsingContext.ArgumentExpressions:
                     return token() === SyntaxKind.DotDotDotToken || isStartOfExpression();
                 case ParsingContext.Parameters:
-                    return isStartOfParameter();
+                    return isStartOfParameter(/*isJSDocParameter*/ false);
+                case ParsingContext.JSDocParameters:
+                    return isStartOfParameter(/*isJSDocParameter*/ true);
                 case ParsingContext.TypeArguments:
                 case ParsingContext.TupleElementTypes:
                     return token() === SyntaxKind.CommaToken || isStartOfType();
@@ -1612,6 +1614,7 @@ namespace ts {
                 case ParsingContext.TupleElementTypes:
                 case ParsingContext.ArrayBindingElements:
                     return token() === SyntaxKind.CloseBracketToken;
+                case ParsingContext.JSDocParameters:
                 case ParsingContext.Parameters:
                 case ParsingContext.RestProperties:
                     // Tokens other than ')' and ']' (the latter for index signatures) are here for better error recovery
@@ -1795,6 +1798,7 @@ namespace ts {
                 case ParsingContext.VariableDeclarations:
                     return isReusableVariableDeclaration(node);
 
+                case ParsingContext.JSDocParameters:
                 case ParsingContext.Parameters:
                     return isReusableParameter(node);
 
@@ -2009,6 +2013,7 @@ namespace ts {
                 case ParsingContext.ArgumentExpressions: return Diagnostics.Argument_expression_expected;
                 case ParsingContext.ObjectLiteralMembers: return Diagnostics.Property_assignment_expected;
                 case ParsingContext.ArrayLiteralMembers: return Diagnostics.Expression_or_comma_expected;
+                case ParsingContext.JSDocParameters: return Diagnostics.Parameter_declaration_expected;
                 case ParsingContext.Parameters: return Diagnostics.Parameter_declaration_expected;
                 case ParsingContext.TypeParameters: return Diagnostics.Type_parameter_declaration_expected;
                 case ParsingContext.TypeArguments: return Diagnostics.Type_argument_expected;
@@ -2430,12 +2435,12 @@ namespace ts {
             return undefined;
         }
 
-        function isStartOfParameter(): boolean {
+        function isStartOfParameter(isJSDocParameter: boolean): boolean {
             return token() === SyntaxKind.DotDotDotToken ||
                 isIdentifierOrPattern() ||
                 isModifierKind(token()) ||
                 token() === SyntaxKind.AtToken ||
-                isStartOfType(/*inStartOfParameter*/ true);
+                isStartOfType(/*inStartOfParameter*/ !isJSDocParameter);
         }
 
         function parseParameter(): ParameterDeclaration {
@@ -2534,7 +2539,9 @@ namespace ts {
             setYieldContext(!!(flags & SignatureFlags.Yield));
             setAwaitContext(!!(flags & SignatureFlags.Await));
 
-            signature.parameters = parseDelimitedList(ParsingContext.Parameters, flags & SignatureFlags.JSDoc ? parseJSDocParameter : parseParameter);
+            signature.parameters = flags & SignatureFlags.JSDoc ?
+                parseDelimitedList(ParsingContext.JSDocParameters, parseJSDocParameter) :
+                parseDelimitedList(ParsingContext.Parameters, parseParameter);
 
             setYieldContext(savedYieldContext);
             setAwaitContext(savedAwaitContext);
@@ -2973,7 +2980,7 @@ namespace ts {
 
         function isStartOfParenthesizedOrFunctionType() {
             nextToken();
-            return token() === SyntaxKind.CloseParenToken || isStartOfParameter() || isStartOfType();
+            return token() === SyntaxKind.CloseParenToken || isStartOfParameter(/*isJSDocParameter*/ false) || isStartOfType();
         }
 
         function parsePostfixTypeOrHigher(): TypeNode {
@@ -6255,6 +6262,7 @@ namespace ts {
             JsxChildren,               // Things between opening and closing JSX tags
             ArrayLiteralMembers,       // Members in array literal
             Parameters,                // Parameters in parameter list
+            JSDocParameters,           // JSDoc parameters in parameter list of JSDoc function type
             RestProperties,            // Property names in a rest type list
             TypeParameters,            // Type parameters in type parameter list
             TypeArguments,             // Type arguments in type argument list

--- a/src/compiler/parser.ts
+++ b/src/compiler/parser.ts
@@ -2967,6 +2967,8 @@ namespace ts {
                 case SyntaxKind.InferKeyword:
                 case SyntaxKind.ImportKeyword:
                     return true;
+                case SyntaxKind.FunctionKeyword:
+                    return !inStartOfParameter;
                 case SyntaxKind.MinusToken:
                     return !inStartOfParameter && lookAhead(nextTokenIsNumericLiteral);
                 case SyntaxKind.OpenParenToken:

--- a/tests/baselines/reference/jsdocParseHigherOrderFunction.symbols
+++ b/tests/baselines/reference/jsdocParseHigherOrderFunction.symbols
@@ -1,0 +1,9 @@
+=== tests/cases/conformance/jsdoc/paren.js ===
+/** @type {function((string), function((string)): string): string} */
+var x = s => s.toString()
+>x : Symbol(x, Decl(paren.js, 1, 3))
+>s : Symbol(s, Decl(paren.js, 1, 7))
+>s.toString : Symbol(String.toString, Decl(lib.es5.d.ts, --, --))
+>s : Symbol(s, Decl(paren.js, 1, 7))
+>toString : Symbol(String.toString, Decl(lib.es5.d.ts, --, --))
+

--- a/tests/baselines/reference/jsdocParseHigherOrderFunction.symbols
+++ b/tests/baselines/reference/jsdocParseHigherOrderFunction.symbols
@@ -1,9 +1,9 @@
 === tests/cases/conformance/jsdoc/paren.js ===
 /** @type {function((string), function((string)): string): string} */
-var x = s => s.toString()
+var x = (s, id) => id(s)
 >x : Symbol(x, Decl(paren.js, 1, 3))
->s : Symbol(s, Decl(paren.js, 1, 7))
->s.toString : Symbol(String.toString, Decl(lib.es5.d.ts, --, --))
->s : Symbol(s, Decl(paren.js, 1, 7))
->toString : Symbol(String.toString, Decl(lib.es5.d.ts, --, --))
+>s : Symbol(s, Decl(paren.js, 1, 9))
+>id : Symbol(id, Decl(paren.js, 1, 11))
+>id : Symbol(id, Decl(paren.js, 1, 11))
+>s : Symbol(s, Decl(paren.js, 1, 9))
 

--- a/tests/baselines/reference/jsdocParseHigherOrderFunction.types
+++ b/tests/baselines/reference/jsdocParseHigherOrderFunction.types
@@ -1,0 +1,11 @@
+=== tests/cases/conformance/jsdoc/paren.js ===
+/** @type {function((string), function((string)): string): string} */
+var x = s => s.toString()
+>x : (arg0: string, arg1: (arg0: string) => string) => string
+>s => s.toString() : (s: string) => string
+>s : string
+>s.toString() : string
+>s.toString : () => string
+>s : string
+>toString : () => string
+

--- a/tests/baselines/reference/jsdocParseHigherOrderFunction.types
+++ b/tests/baselines/reference/jsdocParseHigherOrderFunction.types
@@ -1,11 +1,11 @@
 === tests/cases/conformance/jsdoc/paren.js ===
 /** @type {function((string), function((string)): string): string} */
-var x = s => s.toString()
+var x = (s, id) => id(s)
 >x : (arg0: string, arg1: (arg0: string) => string) => string
->s => s.toString() : (s: string) => string
+>(s, id) => id(s) : (s: string, id: (arg0: string) => string) => string
 >s : string
->s.toString() : string
->s.toString : () => string
+>id : (arg0: string) => string
+>id(s) : string
+>id : (arg0: string) => string
 >s : string
->toString : () => string
 

--- a/tests/baselines/reference/jsdocParseParenthesizedJSDocParameter.symbols
+++ b/tests/baselines/reference/jsdocParseParenthesizedJSDocParameter.symbols
@@ -1,0 +1,9 @@
+=== tests/cases/conformance/jsdoc/paren.js ===
+/** @type {function((string)): string} */
+var x = s => s.toString()
+>x : Symbol(x, Decl(paren.js, 1, 3))
+>s : Symbol(s, Decl(paren.js, 1, 7))
+>s.toString : Symbol(String.toString, Decl(lib.es5.d.ts, --, --))
+>s : Symbol(s, Decl(paren.js, 1, 7))
+>toString : Symbol(String.toString, Decl(lib.es5.d.ts, --, --))
+

--- a/tests/baselines/reference/jsdocParseParenthesizedJSDocParameter.types
+++ b/tests/baselines/reference/jsdocParseParenthesizedJSDocParameter.types
@@ -1,0 +1,11 @@
+=== tests/cases/conformance/jsdoc/paren.js ===
+/** @type {function((string)): string} */
+var x = s => s.toString()
+>x : (arg0: string) => string
+>s => s.toString() : (s: string) => string
+>s : string
+>s.toString() : string
+>s.toString : () => string
+>s : string
+>toString : () => string
+

--- a/tests/cases/conformance/jsdoc/jsdocParseHigherOrderFunction.ts
+++ b/tests/cases/conformance/jsdoc/jsdocParseHigherOrderFunction.ts
@@ -1,0 +1,7 @@
+// @noemit: true
+// @allowjs: true
+// @checkjs: true
+// @strict: true
+// @Filename: paren.js
+/** @type {function((string), function((string)): string): string} */
+var x = s => s.toString()

--- a/tests/cases/conformance/jsdoc/jsdocParseHigherOrderFunction.ts
+++ b/tests/cases/conformance/jsdoc/jsdocParseHigherOrderFunction.ts
@@ -4,4 +4,4 @@
 // @strict: true
 // @Filename: paren.js
 /** @type {function((string), function((string)): string): string} */
-var x = s => s.toString()
+var x = (s, id) => id(s)

--- a/tests/cases/conformance/jsdoc/jsdocParseParenthesizedJSDocParameter.ts
+++ b/tests/cases/conformance/jsdoc/jsdocParseParenthesizedJSDocParameter.ts
@@ -1,0 +1,7 @@
+// @noemit: true
+// @allowjs: true
+// @checkjs: true
+// @strict: true
+// @Filename: paren.js
+/** @type {function((string)): string} */
+var x = s => s.toString()

